services/src/main/java/org/keycloak/jose/jws/DefaultTokenManager.java

```diff
@@ -142,4 +157,94 @@ private String getSignatureAlgorithm(String clientAttribute) {
         return DEFAULT_ALGORITHM_NAME;
     }
 
+    @Override
+    public String encodeAndEncrypt(Token token) {
+        String encodedToken = encode(token);
+        if (isTokenEncryptRequired(token.getCategory())) {
+            encodedToken = getEncryptedToken(token.getCategory(), encodedToken);
+        }
+        return encodedToken;
+    }
+
+    private boolean isTokenEncryptRequired(TokenCategory category) {
+        if (cekManagementAlgorithm(category) == null) return false;
+        if (encryptAlgorithm(category) == null) return false;
+        return true;
+    }
+
+    private String getEncryptedToken(TokenCategory category, String encodedToken) {
+        String encryptedToken = null;
+
+        String algAlgorithm = cekManagementAlgorithm(category);
+        String encAlgorithm = encryptAlgorithm(category);
+
+        CekManagementProvider cekManagementProvider = session.getProvider(CekManagementProvider.class, algAlgorithm);
+        JWEAlgorithmProvider jweAlgorithmProvider = cekManagementProvider.jweAlgorithmProvider();
+
+        ContentEncryptionProvider contentEncryptionProvider = session.getProvider(ContentEncryptionProvider.class, encAlgorithm);
+        JWEEncryptionProvider jweEncryptionProvider = contentEncryptionProvider.jweEncryptionProvider();
+
+        ClientModel client = session.getContext().getClient();
+
+        //KeyWrapper keyWrapper = getEncryptionKekWrapper(algAlgorithm, client);
+        KeyWrapper keyWrapper = PublicKeyStorageManager.getClientPublicKeyWrapper(session, client, JWK.Use.ENCRYPTION, algAlgorithm);
+        if (keyWrapper == null) {
+            throw new RuntimeException("can not get encryption KEK");
+        }
+        Key encryptionKek = keyWrapper.getPublicKey();
+        String encryptionKekId = keyWrapper.getKid();
+        try {
+            encryptedToken = TokenUtil.jweKeyEncryptionEncode(encryptionKek, encodedToken.getBytes("UTF-8"), algAlgorithm, encAlgorithm, encryptionKekId, jweAlgorithmProvider, jweEncryptionProvider);
+        } catch (JWEException | UnsupportedEncodingException e) {
+            throw new RuntimeException(e);
+        }
+        return encryptedToken;
+    }
+
+    private KeyWrapper getEncryptionKekWrapper(String algAlgorithm, ClientModel client) {
```

Comment:
If I am not mistaken, this private method is not used now and can be removed? This will allow to remove PublicKeyStorageManager.getClientPublicKeyWrapper(session, client, JWK.Use.ENCRYPTION) as well as I mentioned in the comment above.

Reply:
Yes, this method has already not been used. I'll remove it.